Indented syntax improvements

jamesnw opened the pull request:
[Rendered Proposal ](https://github.com/oddbird/sass/blob/indented-syntax-improvements/proposal/indented-syntax-improvements.md)

jamesnw commented:
@nex3 This is ready for a high level review. There are missing links, and some areas that need cleanup and expansion, but I wanted to test a few specific concepts for soundness before I continue:

1. The [`Parsing text as Sass`](https://github.com/oddbird/sass/blob/indented-syntax-improvements/proposal/indented-syntax-improvements.md#parsing-text-as-sass) procedure is the conceptual home where I think we want to insert this logic, but I omitted portions orthogonal to this proposal, in other words, actually building the AST.  Is there a better way to frame this?
2. This introduces a dubious `pseudoproduction` concept in [Productions of StatementMayEnd](https://github.com/oddbird/sass/blob/indented-syntax-improvements/proposal/indented-syntax-improvements.md#productions-of-statementmayend). Essentially, I want to define where statements may end. This is not part of the text, so it should not be included in the BNF syntax, but I do want to consume it in the procedures. Is there a better way of defining the exact context where statements may end?

I'd welcome other review, but I expect your feedback on those 2 questions will lead to changes.

nex3 commented on proposal/indented-syntax-improvements.md:
```diff
@@ -0,0 +1,311 @@
+
+<x><pre>
+**NewLine**        ::= LineFeed | CarriageReturn | FormFeed
+**StatementEnd**   ::= NewLine
```
Rather than having `StatementEnd` be its own production, which raises the awkward question of "what's the `StatementEnd` for a statement at the end of a block or document", it would probably be cleaner to define a `StatementSequence` production that consumes all the statements at a given level (either in a block or at the top level of a stylesheet). This would contain a definition of what separates those statements, but it would also be flexible enough to say that a separator isn't necessary at the end of a block in SCSS.

nex3 commented on proposal/indented-syntax-improvements.md:
```diff
@@ -0,0 +1,311 @@
+<x><pre>
+**NewLine**        ::= LineFeed | CarriageReturn | FormFeed
+**StatementEnd**   ::= NewLine
+**BlockStart**     ::= Indent
+**BlockEnd**       ::= Dedent
```
At some point, these will need to more specifically define how to consume a block start and end in terms of the actual sequence of characters in the file.

nex3 reviewed (commented):
I think my biggest suggestion is that, rather than including the existing syntax in the proposal for the new syntax, it should go in the `spec/` directory. It's essentially a specification of how Sass behaves today. Building out that existing structure will both give you a clearer understanding of how the current behavior works and make the new behavior proposal a lot cleaner.

1. In the BNF, I would probably refer to `StatementWithoutBlock` and `StatementWithBlock` productions without providing explicit definitions, then in the prose say "parse the statement according to its specific definition".

2. I think this is something that will be easier once there's an existing well-defined Sass parser spec. I definitely don't think you need this `StatementMayEnd` machinery—arguably it's the current behavior that's _more_ complicated, because it effectively allows a statement to end even where a `;` wouldn't be valid in SCSS. The main difference with the new parse is that in Sass, if a newline appears anywhere a `;` can't go, it doesn't end the statement.

jamesnw commented:
@nex3 This is ready for another mid-process checkin.

The majority of my focus has been on a new [Statement spec ](https://github.com/oddbird/sass/blob/indented-syntax-improvements/spec/statement.md#consuming-a-declaration-value) to define existing behavior. I went down some rabbit trails of trying to specify behavior exactly based on the code, but that section can be ignored for this (and will likely be removed).

In the [Consuming a Statement in the [scss|indented] syntax](https://github.com/oddbird/sass/blob/indented-syntax-improvements/spec/statement.md#consuming-a-statement-in-the-scss-syntax) procedures, I've defined how statements end. However, I haven't defined where statements _can_ end, or more precisely, where semicolons _can_ go, so that newlines elsewhere don't end statements in the indented format. 

Instead of a generic line end algorithm for statements, do I need to fully specify each expression type that could be in a `Statement Without Block`? (variable declarations, `@warn`, `@use`, some property declarations, some CSS at rules, etc). In `stylesheet.dart`, I'm seeing statement ending logic in `almostAnyValue`, `_interpolatedDeclarationValue`, and `_declarationOrBuffer`, but haven't explored enough to see if there's a straightforward heuristic I could specify for all statements.

In testing out the Statement Sequence model, in the `style-rules` spec, I've added a [Procedures](https://github.com/oddbird/sass/blob/indented-syntax-improvements/spec/style-rules.md#procedures) section that specifies how to consume a Style Rule. It seems to parallel the Semantics section, and I wanted to check if it is valid to have both.

For this proposal, do I need to specify the statement routing? For instance, if it starts with `//` consume a Silent Comment, if it starts with `/*` consume a Loud Comment, if it starts with `$` consume a Variable Declaration, etc?

In general, my goal is to document the existing spec that I need to alter in the proposal, but I'm not there yet. Am I headed in the right direction, or where should I head?

nex3 commented:
This is definitely moving in the right direction! My highest-level recommendation is to focus less on prose descriptions of how to parse byte sequences and more on BNF representations of the syntax. You can always add annotations to the BNF for syntax quirks that aren't easily represented in the syntax itself. For example, here's how I'd probably define a sequence of statements in SCSS:

> <x><pre>**ScssStatements** ::= (Statement ';'?¹)* Statement?</pre></x>
>
> 1: This production is mandatory unless the previous `Statement` is a `LoudComment` or `SilentComment`, or its final production is a `Block`.
>
> If a `WhitespaceComment` would be ambiguous with a `Statement` in this rule, parse it preferentially as a `Statement`.

For the indented syntax, you'd probably want to replace the `';'?` with an `IndentSame` production that would, unavoidably, be defined in prose. But I think you _only_ need to define `IndentSame` and `IndentUp` productions in prose, and pretty much everything else (such as the definition of `Block`) can be BNF.

I'd also explicitly define a `Whitespace` production so you have somewhere to make it clear that newlines (and comments that contain newlines) don't currently count as whitespace for the indented syntax. You don't actually need to *use* this production anywhere, though—eventually we should add an explicit specification of how to parse a document based on a BNF that includes "parse whitespace between all productions" but that's not necessary here.

---

> However, I haven't defined where statements _can_ end, or more precisely, where semicolons _can_ go, so that newlines elsewhere don't end statements in the indented format.

This is fine; the productions for each statement make it clear where they can end.

> Instead of a generic line end algorithm for statements, do I need to fully specify each expression type that could be in a `Statement Without Block`?

No, you don't need to fully specify each type. I don't think you even need to split up `StatementWithBlock` and `StatementWithoutBlock` (especially since, as you note, some statement types straddle the two). You can just assume that there's a

<x><pre>**Statement** ::= AtRootRule | ContentRule | DebugRule | ...</pre></x>

definition that we'll write down eventually.

That said, for existing block rules that *do* already have a BNF written down (`ForRule`, `FunctionRule`, `IncludeRule`, `MixinRule`, `UnknownAtRule`, `Declaration`) you should probably change their `'{' Statements '}'` to a `Block` production so you can define it separately for SCSS and the indented syntax.

> In testing out the Statement Sequence model, in the `style-rules` spec, I've added a [Procedures](https://github.com/oddbird/sass/blob/indented-syntax-improvements/spec/style-rules.md#procedures) section that specifies how to consume a Style Rule. It seems to parallel the Semantics section, and I wanted to check if it is valid to have both.

I'm a little confused—I don't see a Semantics section in the current statement spec.

> For this proposal, do I need to specify the statement routing? For instance, if it starts with `//` consume a Silent Comment, if it starts with `/*` consume a Loud Comment, if it starts with `$` consume a Variable Declaration, etc?

No, this is what BNF is for. It lets you say "these are the options, choose whichever one fits at each juncture". The details of how to do that given a byte stream are up to the implementation; as long as the spec unambiguously indicates what the correct behavior is, it doesn't need to go into any detail about _how_ to do it. It's just that sometimes "how" is the easiest way to be unambiguous.

> In general, my goal is to document the existing spec that I need to alter in the proposal, but I'm not there yet. Am I headed in the right direction, or where should I head?

I think you're generally going in the right way, although again focusing more on BNF is a good idea. I'd also recommend defining a term "current indentation" to track the indentation level while parsing an indented syntax document, and using that to define `IndentSame` and `IndentUp`. This is similar to what you're doing now with `indentationLevel`, but specifying it as a parameter doesn't necessarily make much sense in the context of a BNF syntax description.

jamesnw commented:
Thanks! This is helpful!

> > In testing out the Statement Sequence model, in the `style-rules` spec, I've added a [Procedures](https://github.com/oddbird/sass/blob/indented-syntax-improvements/spec/style-rules.md#procedures) section that specifies how to consume a Style Rule. It seems to parallel the Semantics section, and I wanted to check if it is valid to have both.
> 
> I'm a little confused—I don't see a Semantics section in the current statement spec.

I've added this [Procedures](https://github.com/oddbird/sass/blob/indented-syntax-improvements/spec/style-rules.md#procedures) section- but this [Semantics](https://github.com/oddbird/sass/blob/indented-syntax-improvements/spec/style-rules.md#semantics)  section is present in the existing spec. Although if I understand you correctly, I can remove the procedures section and add a BNF production (or assume that there will eventually be a BNF production for Style Rules, outside the scope of this proposal).

nex3 commented:
Ah, I thought you meant a semantics section in your new Statement spec.

Generally, the large-scale structure of a spec matches [that of a proposal](https://github.com/oddbird/sass/blob/indented-syntax-improvements/CONTRIBUTING.md#proposal), so "Syntax" defines how text is converted into a logical representation of the stylesheet (the spec equivalent of an AST), and "Semantics" defines how a stylesheet is executed to produce CSS. "Procedures" is usually used for any behavior that needs to occur multiple times, sort of like function calls.

jamesnw commented:
There are still rough edges, and cleanup, but I think this is ready for another pass, with some questions.

[Rendered Proposal](https://github.com/oddbird/sass/blob/indented-syntax-improvements/proposal/indented-syntax-improvements.md)
[Statement Spec](https://github.com/oddbird/sass/blob/indented-syntax-improvements/spec/statement.md)
[Branch Diff](https://github.com/oddbird/sass/pull/7/files)

1. Is an example of `WhitespaceComment` in your `ScssStatements` production the `here` in `a \* here *\ #id {color: red}`? If so, how can it be ambiguous with a `Statement`? 
2. I assume I'm missing other edge cases- any that jump out?
3. As I currently have it in the proposal, line breaks do not end statements inside curly braces in the indented syntax. Does that make sense? I think it makes sense from a design standpoint to view it as a unidirectional opt-in, but it may not be needed (or even harder) from a code perspective.
4. In testing out the changes on the examples from the proposal, I found some behavior that was unexpected with `@if`, where uninterpretable text is ignored and doesn't throw an error- [example](https://sass-lang.com/playground/#eJw9TssOwjAMu/crfITDkCZu7LJfydYAlfqANJ2GUP+dFiGkSI5jOfY4zsHtLoI2UpLDEW8DzO4KlcKgaBvta4e7y2ijvGvHNYWHZ2X/grvFJGyxlK8Sk4K6Hjhqc/afwJLEsgxC1pV8wXlq12qqMaf8LCQ80PZLj6svlv+dJlM/9JM1xA==?s=L4C3-L4C58). My assumption is that in `@if $a \n and $b`, we would need to end the statement after `$a`, but if we can mirror scss and parse until `IndentMore`, that wouldn't have to cause a new line break. But I also don't want to rely on a bug if that is a bug. 
5. On a related note, I want to confirm the behavior in the flow control rules- I have some @if examples in the [sass-spec PR](https://github.com/oddbird/sass-spec/pull/14/files#diff-53d1524b4a6b99b3a7466bef9f4a7a6535f0b6afb594464e49520d8d53004d77). I set `:ignore_for:` for syntax that wouldn't be supported, and `:todo:` for syntax that will be supported. 
6. You mentioned "comments that contain newlines don't currently count as whitespace for the indented syntax"- is that covered by the whitespace production? 
7. Are there additional things I need to cover in the BNF?
8. I've made lots of changes and cleanup. If I removed prose that would be useful, let me know and I can restore it.
9. I've started to add some test coverage of the indented syntax [here](https://github.com/oddbird/sass-spec/pull/14/files#diff-b72d44dd7b6efb2c5064deab90e37d090a0cdcf8e5db7ce82919a7f2e7e1326b).  Is there other syntax I should make sure has coverage before we change the parsing?

nex3 commented:
I've gone through and answered most of your specific questions. Generally, this is looking good—I think it's worth pulling out the existing-spec portion of this into a separate PR and starting a line-by-line code review of that there, so we can get it landed and squared away as a basis for this proposal. (That's not to say we won't need to tweak it a bit as we iterate on the new proposal, but I think focusing on getting it done first will be valuable.)

> 1. Is an example of `WhitespaceComment` in your `ScssStatements` production the `here` in `a \* here *\ #id {color: red}`? If so, how can it be ambiguous with a `Statement`?

No, an example would be

```scss
/* here */
#id {color: red}
```

The `/* here */` comment could either be parsed as whitespace or as a `LoudComment` statement, but it _should_ be parsed as the latter.

> 2. I assume I'm missing other edge cases- any that jump out?

Edge cases with comments and whitespace or edge cases in general? If the latter, let's wait until the line-by-line review to get into those.

> 3. As I currently have it in the proposal, line breaks do not end statements inside curly braces in the indented syntax. Does that make sense? I think it makes sense from a design standpoint to view it as a unidirectional opt-in, but it may not be needed (or even harder) from a code perspective.

I think it's cleanest if it is, as you say, unidirectional. Everything in curly braces is effectively SCSS.

> 4. In testing out the changes on the examples from the proposal, I found some behavior that was unexpected with `@if`, where uninterpretable text is ignored and doesn't throw an error- [example](https://sass-lang.com/playground/#eJw9TssOwjAMu/crfITDkCZu7LJfydYAlfqANJ2GUP+dFiGkSI5jOfY4zsHtLoI2UpLDEW8DzO4KlcKgaBvta4e7y2ijvGvHNYWHZ2X/grvFJGyxlK8Sk4K6Hjhqc/afwJLEsgxC1pV8wXlq12qqMaf8LCQ80PZLj6svlv+dJlM/9JM1xA==?s=L4C3-L4C58). My assumption is that in `@if $a \n and $b`, we would need to end the statement after `$a`, but if we can mirror scss and parse until `IndentMore`, that wouldn't have to cause a new line break. But I also don't want to rely on a bug if that is a bug.

That text isn't ignored, it's parsed as a space-separated list, which is [truthy](https://sass-lang.com/documentation/values/booleans/#truthiness-and-falsiness) in Sass. You're effectively writing `@if $true and (true this is ...)`.

You're right that in the indented syntax, `@if $a \n and $b` should end the parameter list after `$a`. Basically the idea is that, when it encounters a newline, it asks "is this a valid complete expression?" and if so it's done. We don't want to have to look ahead to see if the expression continues.

> 5. On a related note, I want to confirm the behavior in the flow control rules- I have some @if examples in the [sass-spec PR](https://github.com/oddbird/sass-spec/pull/14/files#diff-53d1524b4a6b99b3a7466bef9f4a7a6535f0b6afb594464e49520d8d53004d77). I set `:ignore_for:` for syntax that wouldn't be supported, and `:todo:` for syntax that will be supported.

I think [this](https://github.com/oddbird/sass-spec/pull/14/files#diff-53d1524b4a6b99b3a7466bef9f4a7a6535f0b6afb594464e49520d8d53004d77R65-R73) should be supported—a bare `@if` without an expression is incomplete, as is `@if $v == up and`.

> 6. You mentioned "comments that contain newlines don't currently count as whitespace for the indented syntax"- is that covered by the whitespace production?

Yes.

> 7. Are there additional things I need to cover in the BNF?
> 8. I've made lots of changes and cleanup. If I removed prose that would be useful, let me know and I can restore it.
> 9. I've started to add some test coverage of the indented syntax [here](https://github.com/oddbird/sass-spec/pull/14/files#diff-b72d44dd7b6efb2c5064deab90e37d090a0cdcf8e5db7ce82919a7f2e7e1326b).  Is there other syntax I should make sure has coverage before we change the parsing?

Let's save these for the line-by-line review as well.

jamesnw commented:
> I think it's worth pulling out the existing-spec portion of this into a separate PR and starting a line-by-line code review of that there, so we can get it landed and squared away as a basis for this proposal.

Great idea- I've opened https://github.com/sass/sass/pull/3969 and https://github.com/sass/sass-spec/pull/2034 as starting places for the existing syntax.

jamesnw commented:
Moved to https://github.com/sass/sass/pull/3973
